src/goose/batch_client.clj

```diff
@@ -0,0 +1,57 @@
+(ns goose.batch-client
+  "Functions for executing a batch of jobs in async manner."
+  (:require [goose.batch :as batch]
+            [goose.broker :as b]
+            [goose.defaults :as d]
+            [goose.job :as job]
+            [goose.retry :as retry]))
+
+(defn- prepare-job
+  [{:keys [queue retry-opts]}
+   execute-fn-sym
+   batch
+   args]
+  (let [retry-opts (retry/prefix-queue-if-present retry-opts)
+        ready-queue (d/prefix-queue queue)
+        job (job/new execute-fn-sym args queue ready-queue retry-opts)]
+    (batch/set-meta batch job)))
+
+(defn- enqueue-batch
+  [broker jobs]
+  (let [batch (batch/new)]
+    (->> (map (fn [{:keys [opts execute-fn-sym args]}]
+                (prepare-job opts execute-fn-sym batch args))
+              jobs)
+      (b/enqueue-batch broker (:id batch)))))
+
+(defn- parse-jobs
+  [opts execute-fn-sym args]
+  (assoc {} :opts opts :execute-fn-sym execute-fn-sym :args args))
+
+(defn perform-batch
+  "Enqueues a Batch of jobs for async execution.
+
+  [broker opts execute-fn-sym job-args]
+   `broker`         : The broker implementation to execute the batch on.
+
+   `opts`           : Map of `:queue` & `:retry-opts`.\\
+   Example          : [[default-opts]]
+
+   `execute-fn-sym` : A fully-qualified function symbol called by worker.\\
+   Example          : ```my-fn`, ```ns-alias/my-fn`, `'fully-qualified-ns/my-fn`
+
+   `args`           : Coll of n variadic values provided in given order when invoking `execute-fn-sym` n times.\\
+    Given values must be serializable by `ptaoussanis/nippy`.
+
+    [broker jobs]
+    `broker`         : The broker implementation to execute the batch on.
+
+    `jobs`           : Map of `:opts`, `:execute-fn-sym` & `args`"
+  ([broker opts execute-fn-sym job-args]
+   (->>
+     (map (fn [args]
+            (parse-jobs opts execute-fn-sym args))
+          job-args)
+     (enqueue-batch broker)))
+  ([broker jobs]
+   (enqueue-batch broker jobs)))
```

Comment:
I'm assuming this will be the API of perform-batch, so have a few comments on that.  This discussion might have multiple back-and-forth. Can do it on slack or over a call.

So, with the current API, users will have to perform a batch like this:
```clojure
; Batch of Jobs with different execute-fn-sym, queue & retry-opts
(let [job1 (goose.job/new `my-fn1 '(:one :args) "queue-1" "goose:queue-1" {:retry :opts})
      job2 (goose.job/new `my-fn2 '(:two :args) "queue-2" "goose:queue-2" {:retry :opts})]
  (perform-batch broker (list job1 job2)))

; Batch of Jobs with common execute-fn-sym, queue & retry-opts
(let [opts {:queue "queue-1"
            :retry-opts {:retry :opts}}
      job-args (list '(:one :args) '(:two :args))]
  (perform-batch broker opts `my-fn job-args))
```

Please interrupt me if my understanding of the API is incorrect here. This usage is bad IMO because of 2 reasons:
1. User-experience is bad when they have to give args in a list instead of in a variadic manner like they'd do to a synchronous function
2. We are exposing `job.new` which is an internal implementation detail of Goose. We're even not generating docstrings for that namespace.

Reply:
When I thought about the API design, it kinda occurred to me that a user mostly uses batching for the same function, and a collection of sequential args.

If we restrict users to function, queue & retry-opts, that takes away flexibility but reduces syntax complexity drastically. IMO, that's a fair trade-off to make.

This is my idea of batching interface, to keep it functional & in-line with other APIs of Goose.
```clojure
(defn accumulate-batch-args
  [list & args]
  (conj list args))

(defn perform-batch
  [opts execute-fn-sym args]
  (let
     ; iterate over args & construct jobs
    [jobs (map job/new execute-fn-sym (:queue opts) args)
     ; create a batch
     batch (batch/new)]
    ; enqueue them to broker in 1 transaction
    (enqueue-batch (:broker opts) batch jobs)))

(let [my-args '(1 2 3 4 5)
      batch-args (reduce accumulate-batch-args '() my-args)
      opts {:queue "queue-1"
            :retry-opts {:retry :opts}
            :broker "redis"}]
  (perform-batch opts `my-fn batch-args))
```

Reply:
@olttwa based on our conversation, I've update `perform-batch` to accept common `opts` and `execute-fn-sym`. 
Do you think we should also have the `accumulate-batch-args` helper function in the client?

Reply:
@siripr4 helper function can go to batch namespace & be called `batch/accumulate-args`.

Reply:
We would be exposing the batch namespace to users. Is that okay?

I'm inclining towards adding this to `goose.client` instead. Similar to how we currently have the `default-opts` helper function.

Reply:
Makes sense. Let's add to client namespace instead. Batch can stay an internal namespace for the moment.

Reply:
Moved `accumulate-batch-args` to client.